In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime

import warnings
warnings.filterwarnings('ignore')

## 1. PREPARE DATA

### 1.1. Read CustomerDemographic Table

In [2]:
df_dem = pd.read_excel('./data/KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=3, skiprows=[0], usecols='A:M', engine='openpyxl')
df_dem.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [3]:
df_dem.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [4]:
print('The CustomerDemographic Table has', df_dem.shape[0], 'rows and', df_dem.shape[1], 'columns.')

The CustomerDemographic Table has 4000 rows and 13 columns.


### 1.2. Read CustomerAddress Table

In [5]:
df_add = pd.read_excel('./data/KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=4, skiprows=[0], usecols='A:F', engine='openpyxl')
df_add.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [6]:
df_add.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [7]:
print('The CustomerAddress Table has', df_add.shape[0], 'rows and', df_add.shape[1], 'columns.')

The CustomerAddress Table has 3999 rows and 6 columns.


### 1.3. Read Transactions Table

In [10]:
df_trans = pd.read_excel('./data/KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=1, skiprows=[0], usecols='A:M', engine='openpyxl')
df_trans.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [11]:
df_trans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [12]:
print('The Transactions Table has', df_trans.shape[0], 'rows and', df_trans.shape[1], 'columns.')

The Transactions Table has 20000 rows and 13 columns.


### 1.4. Read NewCustomerList Table

In [13]:
df_newlist = pd.read_excel('./data/KPMG_VI_New_raw_data_update_final.xlsx', sheet_name=2, skiprows=[0], usecols='A:W', engine='openpyxl')
df_newlist.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        1.03      1.2875     1.609375     1.367969   
1                  11        0.86      0.8600     1.075000     0.913750   
2                   5        0.52      0.5200     0.520000     0.520000   
3                   1        0.88      1.1000     1.100000     1.100000   
4                   9        0.54      0.5400     0.675000     0.675000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [14]:
df_newlist.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [15]:
print('The NewCustomerList Table has', df_newlist.shape[0], 'rows and', df_newlist.shape[1], 'columns.')

The NewCustomerList Table has 1000 rows and 23 columns.


## 2. PROCESS DATA

### 2.1. Check data in 3 tables

**CustomerDemographic**

In [17]:
# Check the data for duplicates
print('Shape before dropping duplicates', df_dem.shape)
df_dem = df_dem.drop_duplicates()
print('Shape after dropping duplicates', df_dem.shape)

Shape before dropping duplicates (4000, 13)
Shape after dropping duplicates (4000, 13)


CustomerDemographic Table has no duplicates

In [18]:
# Check the data for missing values
print(np.sum(df_dem.isnull()), '\n')

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64 



 There are lots of missing data records in CustomerDemographic Table

**CustomerAddress**

In [19]:
# Check the data for duplicates
print('Shape before dropping duplicates', df_add.shape)
df_add = df_add.drop_duplicates()
print('Shape after dropping duplicates', df_add.shape)

Shape before dropping duplicates (3999, 6)
Shape after dropping duplicates (3999, 6)


CustomerAddress Table has no duplicates

In [20]:
# Check the data for missing values
print(np.sum(df_add.isnull()), '\n')

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64 



 There are no missing data records in CustomerAddress Table

**Transactions**

In [21]:
# Check the data for duplicates
print('Shape before dropping duplicates', df_trans.shape)
df_trans = df_trans.drop_duplicates()
print('Shape after dropping duplicates', df_trans.shape)

Shape before dropping duplicates (20000, 13)
Shape after dropping duplicates (20000, 13)


Transactions Table has no duplicates

In [22]:
# Check the data for missing values
print(np.sum(df_trans.isnull()), '\n')

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64 



 There are lots of missing data records in Transactions Table

### 2.2. Merge 3 tables into 1 table

In [23]:
df_1 = pd.merge(df_dem, df_add, how='outer', on=['customer_id'])

In [27]:
df_dem.shape

(4000, 13)

In [28]:
df_add.shape

(3999, 6)

In [26]:
df_1.shape

(4003, 18)

In [30]:
df_all = pd.merge(df_trans, df_1, how='outer', on=['customer_id'])

In [32]:
df_trans.shape

(20000, 13)

In [31]:
df_all.shape

(20510, 30)

In [33]:
df_1.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure', 'address', 'postcode', 'state',
       'country', 'property_valuation'],
      dtype='object')

In [34]:
df_trans.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

In [35]:
df_all.head(3)

transaction_id  product_id  customer_id transaction_date  online_order  \
0             1.0         2.0         2950       2017-02-25           0.0   
1         11065.0         1.0         2950       2017-10-16           0.0   
2         18923.0        62.0         2950       2017-04-26           0.0   

  order_status           brand product_line product_class product_size  ...  \
0     Approved           Solex     Standard        medium       medium  ...   
1     Approved  Giant Bicycles     Standard        medium       medium  ...   
2     Approved           Solex     Standard        medium       medium  ...   

   wealth_segment  deceased_indicator           default owns_car tenure  \
0   Mass Customer                   N  ã»(ï¿£âï¿£)ã»:*:      Yes   10.0   
1   Mass Customer                   N  ã»(ï¿£âï¿£)ã»:*:      Yes   10.0   
2   Mass Customer                   N  ã»(ï¿£âï¿£)ã»:*:      Yes   10.0   

             address  postcode state    country property_valuation  
0  984 Hoepker Court    3064.0   VIC  Australia                6.0  
1  984 Hoepker Court    3064.0   VIC  Australia                6.0  
2  984 Hoepker Court    3064.0   VIC  Australia                6.0  

[3 rows x 30 columns]

In [36]:
# Check the data for missing values
print(np.sum(df_all.isnull()), '\n')

transaction_id                          510
product_id                              510
customer_id                               0
transaction_date                        510
online_order                            870
order_status                            510
brand                                   707
product_line                            707
product_class                           707
product_size                            707
list_price                              510
standard_cost                           707
product_first_sold_date                 707
first_name                                6
last_name                               661
gender                                    6
past_3_years_bike_related_purchases       6
DOB                                     463
job_title                              2483
job_industry_category                  3330
wealth_segment                            6
deceased_indicator                        6
default                         

There are **510 customers without having any transactions** => Need to be removed 510 records where 'transaction_id' is empty

### 2.3. The table for analyzing

In [55]:
df_all = df_all.loc[~df_all['transaction_id'].isnull()]

In [56]:
df_all.shape

(20000, 30)

In [57]:
df_all.tail(3)

transaction_id  product_id  customer_id transaction_date  online_order  \
19997         18462.0        80.0         2789       2017-06-20           0.0   
19998         17981.0        69.0         3446       2017-12-26           1.0   
19999         18165.0        86.0         3446       2017-12-03           0.0   

      order_status           brand product_line product_class product_size  \
19997     Approved      OHM Cycles      Touring           low       medium   
19998     Approved  Giant Bicycles         Road        medium       medium   
19999     Approved      OHM Cycles     Standard        medium       medium   

       ...     wealth_segment  deceased_indicator            default owns_car  \
19997  ...  Affluent Customer                   N  ãã¼ãã£ã¼ã¸è¡ããªãã      Yes   
19998  ...      Mass Customer                   N                NaN       No   
19999  ...      Mass Customer                   N                NaN       No   

      tenure         address  postcode state    country property_valuation  
19997    7.0   724 West Park    2112.0   NSW  Australia               11.0  
19998   14.0  8 Becker Drive    4868.0   QLD  Australia                4.0  
19999   14.0  8 Becker Drive    4868.0   QLD  Australia                4.0  

[3 rows x 30 columns]

In [58]:
print('The Table has', df_all.shape[0], 'rows and', df_all.shape[1], 'columns.')

The Table has 20000 rows and 30 columns.


### 2.4. Sort data